### PR TITLE
First pass at using PRAW streams and celery task

### PR DESCRIPTION
**Pull request by narthollis**
Firstly, This PR is mostly untested and will very likely need work to get it operating 100%. I am happy to be to answer most any questions relating to it.

Addresses #12, #2, and #3 - this should also help (but not completely) address #5 and #1

It should also address #4 through increased celery worker count and being able to shard workers onto additional servers - that being said, I didn't bother to read up on Reddit's API limits so more work may be required here.

Mostly this PR is a refactor of the existing code base to play nicely with celery. In the process I also converted the 2 ingestion entry points (new submissions and new inbox messages) to use PRAW streams.

If my understanding is correct `hfysubs.submissions.handle_submission_stream` and `hfysubs.inbox.handle_inbox_stream` should both block on while waiting for new messages to be provided to them by the PRAW stream API.

These handlers then trigger a celery task that performs any necessary database operations, and then call a celery task in `hfysubs.tasks.reddit_writer`

The tasks in  `hfysubs.tasks.reddit_writer` are set to be rate limited by celery and should thereby help reduce the case of running into Reddit rate limit errors. The writer tasks are also setup to retry up to 5 times before failing. When they fail the last time they will raise an uncaught error into Celery which will be logged per-celery defaults.

I have also updated `start.sh` to (hopefully) include appropriate defaults  for starting the stream handlers and celery worker queues.

I hope this is helpful in getting HFYSubs back into good health - it's one of the most useful bots I interact with regularly!

### CODE REVIEW

File: hfysubs/submissions.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+
+def handle_subscription_stream():
+    subreddit = make_reddit().subreddit(config.SUBREDDIT)
+
+    latest = config.get_latest()
+
+    for submission in subreddit.stream.submissions():
+        check = filter_post(submission)
+        if submission.id != latest and check and submission.author is not None:
```
Comment (TheDarkLordSano) [inline hfysubs/submissions.py]: latest is in here from an old method.  with using the stream we shouldn't need to check this.

latest was the Hex number in the web address to define a specific post. like 6dn2wr defines [this](https://www.reddit.com/r/HFY/comments/6dn2wr/life_with_my_2d_waifu_13_aiverse/)

Comment (TheDarkLordSano) [reply hfysubs/submissions.py]: same with line 12. we should be able to delete that outright.

**Comment by TheDarkLordSano** [comment]
Reddit API is set to timeout at 60 queries/actions per minute.

**Comment by TheDarkLordSano** [comment]
Additionally, The way I'm understanding [subreddit.stream.submissions()](http://praw.readthedocs.io/en/v4.5.1/code_overview/models/subreddit.html#praw.models.Subreddit.stream)  is that it is going to open an infinite loop of checking for new submissions when it is called. I also understand it that when stream doesn't find a new submission or inbox message it will automatically wait before executing again. This interval that it waits grows to a set maximum and resets on finding new.

**Comment by narthollis** [comment]
Regarding the infinite looping behavior of `stream` I am relying on it being so with this code.

The two stream handler methods exist only to handle new stream items and create downstream tasks.
You can see in main.py#8 that i create a sub-process for each stream handler so that they can both run at the same time.

**Comment by TheDarkLordSano** [comment]
Okay understandable. I guess I just don't understand completely the timeout feature of the multiprocessing's .join()

**Comment by narthollis** [comment]
Ah, the timeout on `Process.join` is so that we keep checking both child processes for completion.

`join()` will only succeed when the Process is finished, otherwise it will block indefinitely.

I wanted to set up `main.py` such that if either handler process finished for whatever reason, we would kill the other process and end the parent. I figured this would make it easier to monitor if the handlers were running.

Another option would be split each handler into its own startup, which would then require having the operator monitor both.

**Comment by TheDarkLordSano** [comment]
Okay I believe i'm following you.

Now for my own clarification: When say process_submission(submission) is called it is added to the queue of things to do because it's been marked with @app.task(base=RedditTask)?  Or even  send_message()?

**Comment by narthollis** [comment]
Yeah, the `@app.task` is a decorator from Celery that performs the queueing magic.

It basically intercepts any calls to the decorated function and adds them to a task queue instead of invoking the function directly.

Then in the celery process, it watches these queues and invokes the real function on a free worker as one becomes available.

(Celery also automatically serializes and deserializes the call parameters for us, which is super nice and means we can for the most part treat tasks like any other function. It also supports return values as well, though I have avoided them as it requires more setup on the Celery side of things)

**Pull request merged**